### ISSUE TITLE
VCF to SNPxSNP fast epistasis 

### ISSUE THREAD

**Issue by joneill-code** [chrchang/plink-ng#84]
Hello! I have dst information on drug resistant and susceptible bacteria in the form of VCF. How can I go from my many VCF files to the input required for fast epistasis?

**Comment by chrchang** [comment]
Write a script to run "plink --vcf [name of VCF] --make-bed [prefix for output files]" on all the VCFs, and then use "plink --merge-list" to merge the results.  See https://www.cog-genomics.org/plink/1.9/data#merge for more documentation on the merge function.

**Comment by joneill-code** [comment]
Thank you! I have created the bed files and attempting to merge the list of outputs with --merge-list, and I'm having trouble wrapping my head around the merge documentation. 
I attempt the following:

./plink --merge-list test.*.vcf.gz.bed

(I have 1400 roughly input files, an example of one is: "test.TB1285.vcf.gz.bed"  - replace the isolate ID with an "*" to input the whole list.) 
I get the error: 
"Error: --merge-list only accepts 1 parameter."
Any idea how I can work around this?

**Comment by joneill-code** [comment]
Update: I am having trouble merging the VCFs and have read the documentation thoroughly

Here is the error message-

**joneill@valafar00:/grp/valafar/data/pyseer-gwa/pza/vcfs$ ./plink --bfile --merge-list filenames.txt --make-bed --out mymerged
PLINK v1.90b6.4 64-bit (7 Aug 2018)            www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to mymerged.log.
Options in effect:
  --bfile
  --make-bed
  --merge-list filenames.txt
  --out mymerged

24155 MB RAM detected; reserving 12077 MB for main workspace.
Error: Failed to open test.1-0002.vcf.gz.bed.fam.**

Where filenames.txt is the list of file names. I am unsure why I get an error message when the merge is attempted.

**Comment by chrchang** [comment]
Remove the ".bed" suffix on each line of the --merge-list file.  Whenever a --merge-list line has only one token, plink treats it as the common filename *prefix* of the .bed, .bim, and .fam files.

**Comment by joneill-code** [comment]
Thank you! That worked. Last I ran the bed file through fast-epistasis, I knew this would be an issue. :: How do do I input their respective phenotypes? Last, I get this error with the merged file.

PLINK v1.90b6.4 64-bit (7 Aug 2018)            www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to plink.log.
Options in effect:
  --bfile test
  --fast-epistasis

24155 MB RAM detected; reserving 12077 MB for main workspace.
70282 variants loaded from .bim file.
703 people (0 males, 0 females, 703 ambiguous) loaded from .fam.
Ambiguous sex IDs written to plink.nosex .
Using up to 11 threads (change this with --threads).
Before main variant filters, 703 founders and 0 nonfounders present.
Calculating allele frequencies... done.
70282 variants and 703 people pass filters and QC.
Note: No phenotypes present.
**Error: --fast-epistasis requires at least two cases and two controls.**

Thank you for your time

**Comment by chrchang** [comment]
--pheno is usually the simplest way to do this.

**Comment by joneill-code** [comment]
Last issue with VCF. The output file below does not have the SNP positions and instead are ".". I'm assuming it's because SNP IDS are periods in bacteria while labeled in humans. Is there anyway to override the SNP from the VCFs with their respective position?

**Comment by chrchang** [comment]
--set-missing-var-ids can be used for this.  If you also want to overwrite non-. IDs, use plink 2.0's --set-all-var-ids instead.

**Comment by joneill-code** [comment]
Dr. Chang, using the --epistasis command on 1431 clinical isolates resulted in the following strong chi squared value:

1:1637015:A:G | 730 | 13014 | 0.05609 | **129100000000** | 1 | 1:3054321:A:G
-- | -- | -- | -- | -- | -- | --


However, when I look at the sheer counts of isolates having those two variants,
344 are Resistant and 882 are susceptible out of that 1431 which have those two variants; no visible correlation between phenotype and the two SNPs is seen. Does that make it false positive, or is there something I'm missing with interpreting the output?
Thank you again for your time

**Comment by chrchang** [comment]
This looks like some sort of bug.  Can you send me a .log file and test dataset to replicate this with?

**Comment by chrchang** [comment]
Turned out to be caused by a hole in the logistic regression routine's convergence failure detector, which should be fixed by commit b208252 (latest development build includes this).

**Comment by joneill-code** [comment]
Dr. Chang, I wrote an email. The data is still suspect. Perhaps there is simply no simple epistatic interaction.

**Comment by chrchang** [comment]
Yes, b208252 doesn't fully solve the problem.  Working on a more complete solution now.

**Comment by chrchang** [comment]
Remaining spurious results should be removed as of 2680b99 .

**Comment by Junjun-Xu** [comment]
@joneill-code 
hi, can you tell me how to deal with problem of "Error: --fast-epistasis requires at least two cases and two controls.". I don not know how to set the case and control for my data. I what to detect epistasis between subgenome in a species. So i have the same sample but different chromosome for my data. Thanks a lot.

**Comment by flystar233** [comment]
> @joneill-code
> hi, can you tell me how to deal with problem of "Error: --fast-epistasis requires at least two cases and two controls.". I don not know how to set the case and control for my data. I what to detect epistasis between subgenome in a species. So i have the same sample but different chromosome for my data. Thanks a lot.

准备常规的表型文件（FID IID phenotype）,在plink命令后加 `--allow-no-sex` 可以解决问题。

**Comment by Junjun-Xu** [comment]
非常感谢！我的问题已经解决啦！
